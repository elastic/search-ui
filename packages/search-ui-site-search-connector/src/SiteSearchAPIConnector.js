import adaptRequest from "./requestAdapter";
import adaptResponse from "./responseAdapter";
import request from "./request";

function _get(engineKey, path, params) {
  const query = Object.entries({ engine_key: engineKey, ...params })
    .map(([paramName, paramValue]) => {
      return `${paramName}=${encodeURIComponent(paramValue)}`;
    })
    .join("&");

  return fetch(
    `https://search-api.swiftype.com/api/v1/public/${path}?${query}`,
    {
      method: "GET",
      credentials: "include"
    }
  );
}

class SiteSearchAPIConnector {
  /**
   * @callback next
   * @param {Object} updatedQueryOptions The options to send to the API
   */

  /**
   * @callback hook
   * @param {Object} queryOptions The options that are about to be sent to the API
   * @param {next} next The options that are about to be sent to the API
   */

  /**
   * @typedef Options
   * @param  {string} documentType Document Type found in your Site Search Dashboard
   * @param  {string} engineKey Credential found in your Site Search Dashboard
   * @param  {hook} beforeSearchCall=(queryOptions,next)=>next(queryOptions) A hook to amend query options before the request is sent to the
   *   API in a query on an "onSearch" event.
   * @param  {hook} beforeAutocompleteResultsCall=(queryOptions,next)=>next(queryOptions) A hook to amend query options before the request is sent to the
   *   API in a "results" query on an "onAutocomplete" event.
   */

  /**
   * @param {Options} options
   */
  constructor({
    documentType,
    engineKey,
    beforeSearchCall = (queryOptions, next) => next(queryOptions),
    beforeAutocompleteResultsCall = (queryOptions, next) => next(queryOptions)
  }) {
    this.documentType = documentType;
    this.engineKey = engineKey;
    this.beforeSearchCall = beforeSearchCall;
    this.beforeAutocompleteResultsCall = beforeAutocompleteResultsCall;
    this.request = request.bind(this, engineKey);
    this._get = _get.bind(this, engineKey);
  }

  onResultClick({ query, documentId, tags }) {
    if (tags && tags.length > 0) {
      console.warn(
        "search-ui-site-search-connector: Site Search does not support tags on click"
      );
    }
    this._get("analytics/pc", {
      t: new Date().getTime(),
      q: query,
      doc_id: documentId
    });
  }

  onAutocompleteResultClick({ query, documentId, tags }) {
    if (tags) {
      console.warn(
        "search-ui-site-search-connector: Site Search does not support tags on autocompleteClick"
      );
    }
    this._get("analytics/pas", {
      t: new Date().getTime(),
      q: query,
      doc_id: documentId
    });
  }

  onSearch(state, queryConfig) {
    const options = adaptRequest(state, queryConfig, this.documentType);

    return this.beforeSearchCall(options, newOptions =>
      this.request("POST", "engines/search.json", newOptions).then(json =>
        adaptResponse(json, this.documentType)
      )
    );
  }

  async onAutocomplete({ searchTerm }, queryConfig) {
    if (queryConfig.results) {
      const options = adaptRequest(
        { searchTerm },
        queryConfig.results,
        this.documentType
      );

      return this.beforeAutocompleteResultsCall(options, newOptions =>
        this.request("POST", "engines/suggest.json", newOptions).then(json => ({
          autocompletedResults: adaptResponse(json, this.documentType).results
        }))
      );
    }
    if (queryConfig.suggestions) {
      console.warn(
        "search-ui-site-search-connector: Site Search does support query suggestions on autocomplete"
      );
    }
  }
}

export default SiteSearchAPIConnector;
