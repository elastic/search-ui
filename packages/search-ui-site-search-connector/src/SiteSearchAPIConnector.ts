import adaptRequest from "./requestAdapter";
import adaptResponse from "./responseAdapter";
import request from "./request";
import type { RequestState } from "@elastic/search-ui";
import type {
  SearchQueryHook,
  SiteSearchAPIConnectorParams,
  SiteSearchQueryConfig
} from "./types";

function _get(engineKey: string, path: string, params: Record<string, any>) {
  const query = Object.entries({ engine_key: engineKey, ...params })
    .map(([paramName, paramValue]) => {
      return `${paramName}=${encodeURIComponent(paramValue)}`;
    })
    .join("&");

  return fetch(
    `https://search-api.swiftype.com/api/v1/public/${path}?${query}`,
    {
      method: "GET",
      credentials: "include"
    }
  );
}

class SiteSearchAPIConnector {
  /**
   * @callback next
   * @param {Object} updatedQueryOptions The options to send to the API
   */

  /**
   * @callback hook
   * @param {Object} queryOptions The options that are about to be sent to the API
   * @param {next} next The options that are about to be sent to the API
   */

  /**
   * @typedef Options
   * @param  {string} documentType Document Type found in your Site Search Dashboard
   * @param  {string} engineKey Credential found in your Site Search Dashboard
   * @param  {hook} beforeSearchCall=(queryOptions,next)=>next(queryOptions) A hook to amend query options before the request is sent to the
   *   API in a query on an "onSearch" event.
   * @param  {hook} beforeAutocompleteResultsCall=(queryOptions,next)=>next(queryOptions) A hook to amend query options before the request is sent to the
   *   API in a "results" query on an "onAutocomplete" event.
   */

  /**
   * @param {Options} options
   */

  _get: (path: string, params: Record<string, any>) => Promise<Response>;
  request: (
    method: string,
    path: string,
    params: Record<string, any>
  ) => Promise<any>;
  beforeAutocompleteResultsCall: SearchQueryHook;
  documentType: string;
  engineKey: string;
  beforeSearchCall: SearchQueryHook;

  constructor({
    documentType,
    engineKey,
    beforeSearchCall = (queryOptions, next) => next(queryOptions),
    beforeAutocompleteResultsCall = (queryOptions, next) => next(queryOptions)
  }: SiteSearchAPIConnectorParams) {
    this.documentType = documentType;
    this.engineKey = engineKey;
    this.beforeSearchCall = beforeSearchCall;
    this.beforeAutocompleteResultsCall = beforeAutocompleteResultsCall;
    this.request = request.bind(this, engineKey);
    this._get = _get.bind(this, engineKey);
  }

  onResultClick({
    query,
    documentId,
    tags
  }: {
    query: string;
    documentId: string;
    tags: string[];
  }) {
    if (tags && tags.length > 0) {
      console.warn(
        "search-ui-site-search-connector: Site Search does not support tags on click"
      );
    }
    this._get("analytics/pc", {
      t: new Date().getTime(),
      q: query,
      doc_id: documentId
    });
  }

  onAutocompleteResultClick({
    query,
    documentId,
    tags
  }: {
    query: string;
    documentId: string;
    tags: string[];
  }) {
    if (tags) {
      console.warn(
        "search-ui-site-search-connector: Site Search does not support tags on autocompleteClick"
      );
    }
    this._get("analytics/pas", {
      t: new Date().getTime(),
      q: query,
      doc_id: documentId
    });
  }

  onSearch(requestState: RequestState, queryConfig: SiteSearchQueryConfig) {
    const options = adaptRequest(requestState, queryConfig, this.documentType);

    return this.beforeSearchCall(options, (newOptions) =>
      this.request("POST", "engines/search.json", newOptions).then((json) =>
        adaptResponse(json, this.documentType)
      )
    );
  }

  async onAutocomplete(
    { searchTerm }: { searchTerm: string },
    queryConfig: { results?: SiteSearchQueryConfig; [key: string]: any }
  ) {
    if (queryConfig.results) {
      const options = adaptRequest(
        { searchTerm },
        queryConfig.results,
        this.documentType
      );

      return this.beforeAutocompleteResultsCall(options, (newOptions) =>
        this.request("POST", "engines/suggest.json", newOptions).then(
          (json) => ({
            autocompletedResults: adaptResponse(json, this.documentType).results
          })
        )
      );
    }
    if (queryConfig.suggestions) {
      console.warn(
        "search-ui-site-search-connector: Site Search does support query suggestions on autocomplete"
      );
    }
  }
}

export default SiteSearchAPIConnector;
