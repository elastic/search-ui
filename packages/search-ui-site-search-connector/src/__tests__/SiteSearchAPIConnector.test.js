import SiteSearchAPIConnector from "..";

import exampleAPIResponse from "../../resources/example-response.json";

function fetchResponse(response) {
  return Promise.resolve({
    status: 200,
    json: () => Promise.resolve(response)
  });
}

beforeEach(() => {
  global.Headers = jest.fn();
  global.fetch = jest.fn().mockReturnValue(fetchResponse(exampleAPIResponse));
});

const engineKey = 12345;
const documentType = "national-parks";
const params = {
  documentType,
  engineKey
};

it("can be initialized", () => {
  const connector = new SiteSearchAPIConnector(params);
  expect(connector).toBeInstanceOf(SiteSearchAPIConnector);
});

describe("#onSearch", () => {
  function subject({ beforeSearchCall, state, queryConfig = {} }) {
    const connector = new SiteSearchAPIConnector({
      ...params,
      beforeSearchCall
    });
    return connector.onSearch(state, queryConfig);
  }

  it("will correctly format an API response", async () => {
    const response = await subject({ state: {}, queryConfig: {} });
    expect(response).toMatchSnapshot();
  });

  it("will pass request state through to search endpoint", async () => {
    const state = {
      searchTerm: "searchTerm",
      current: 1,
      resultsPerPage: 10,
      sortDirection: "desc",
      sortField: "name",
      filters: [
        {
          field: "title",
          type: "all",
          values: ["Acadia", "Grand Canyon"]
        },
        {
          field: "world_heritage_site",
          values: ["true"],
          type: "all"
        }
      ]
    };

    await subject({ state });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      page: 1,
      per_page: 10,
      filters: {
        [documentType]: {
          title: {
            type: "and",
            values: ["Acadia", "Grand Canyon"]
          },
          world_heritage_site: {
            type: "and",
            values: ["true"]
          }
        }
      },
      sort_direction: {
        "national-parks": "desc"
      },
      sort_field: {
        "national-parks": "name"
      },
      q: "searchTerm"
    });
  });

  it("will pass queryConfig to search endpoint", async () => {
    const state = {
      searchTerm: "searchTerm"
    };

    const queryConfig = {
      facets: {
        states: {
          type: "value",
          size: 30
        }
      },
      result_fields: {
        title: { raw: {}, snippet: { size: 20, fallback: true } }
      },
      search_fields: {
        title: {},
        description: {},
        states: {}
      }
    };

    await subject({ state, queryConfig });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      facets: {
        "national-parks": ["states"]
      },
      fetch_fields: {
        [documentType]: ["title"]
      },
      search_fields: {
        [documentType]: ["title", "description", "states"]
      },
      highlight_fields: {
        [documentType]: {
          title: { size: 20, fallback: true }
        }
      },
      q: "searchTerm"
    });
  });

  it("will pass request parameter state provided to queryConfig, overriding the same value provided in state", async () => {
    const state = {
      searchTerm: "searchTerm",
      current: 1,
      resultsPerPage: 10,
      sortDirection: "desc",
      sortField: "name",
      filters: [
        {
          field: "title",
          type: "all",
          values: ["Acadia", "Grand Canyon"]
        },
        {
          field: "world_heritage_site",
          values: ["true"],
          type: "all"
        }
      ]
    };

    const queryConfig = {
      current: 2,
      resultsPerPage: 5,
      sortDirection: "asc",
      sortField: "title",
      filters: [
        {
          field: "date_made",
          values: ["yesterday"],
          type: "all"
        }
      ]
    };

    await subject({ state, queryConfig });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      page: 2,
      per_page: 5,
      filters: {
        [documentType]: {
          date_made: {
            type: "and",
            values: ["yesterday"]
          }
        }
      },
      sort_direction: {
        "national-parks": "asc"
      },
      sort_field: {
        "national-parks": "title"
      },
      q: "searchTerm"
    });
  });

  it("will use the beforeSearchCall parameter to to amend option parameters to the search endpoint call", async () => {
    const searchTerm = "searchTerm";

    const queryConfig = {
      sortDirection: "desc",
      sortField: "name",
      resultsPerPage: 5
    };

    const beforeSearchCall = (options, next) => {
      // Remove sort_direction and sort_field
      // eslint-disable-next-line no-unused-vars
      const { sort_direction, sort_field, ...rest } = options;
      return next({
        ...rest,
        // Add group
        group: { field: "title" }
      });
    };

    await subject({
      beforeSearchCall,
      state: { searchTerm },
      queryConfig
    });

    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      per_page: 5,
      engine_key: engineKey,
      q: "searchTerm",
      group: { field: "title" }
    });
  });
});

describe("#onAutocomplete", () => {
  function subject({
    beforeAutocompleteResultsCall,
    state,
    queryConfig = {
      results: {}
    }
  }) {
    const connector = new SiteSearchAPIConnector({
      beforeAutocompleteResultsCall,
      ...params
    });
    return connector.onAutocomplete(state, queryConfig);
  }

  it("will correctly format an API response", async () => {
    const response = await subject({
      state: {},
      queryConfig: {
        results: {}
      }
    });
    expect(response).toMatchSnapshot();
  });

  it("will pass searchTerm from state through to search endpoint", async () => {
    const state = {
      searchTerm: "searchTerm"
    };

    await subject({ state });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      q: "searchTerm"
    });
  });

  it("will pass queryConfig to search endpoint", async () => {
    const state = {
      searchTerm: "searchTerm"
    };

    const queryConfig = {
      results: {
        result_fields: {
          title: { raw: {}, snippet: { size: 20, fallback: true } }
        },
        search_fields: {
          title: {},
          description: {},
          states: {}
        }
      }
    };

    await subject({ state, queryConfig });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      q: "searchTerm",
      fetch_fields: {
        [documentType]: ["title"]
      },
      search_fields: {
        [documentType]: ["title", "description", "states"]
      },
      highlight_fields: {
        [documentType]: {
          title: { size: 20, fallback: true }
        }
      }
    });
  });

  it("will pass request parameter state provided to queryConfig", async () => {
    const state = {
      searchTerm: "searchTerm"
    };

    const queryConfig = {
      results: {
        current: 2,
        resultsPerPage: 5,
        filters: [
          {
            field: "world_heritage_site",
            values: ["true"],
            type: "all"
          }
        ],
        sortDirection: "desc",
        sortField: "name"
      }
    };

    await subject({ state, queryConfig });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      q: "searchTerm",
      page: 2,
      per_page: 5,
      filters: {
        [documentType]: {
          world_heritage_site: {
            type: "and",
            values: ["true"]
          }
        }
      },
      sort_direction: {
        "national-parks": "desc"
      },
      sort_field: {
        "national-parks": "name"
      }
    });
  });

  it("will use the beforeAutocompleteResultsCall parameter to amend option parameters to the search endpoint call", async () => {
    const state = {
      searchTerm: "searchTerm"
    };

    const queryConfig = {
      results: {
        sortDirection: "desc",
        sortField: "name",
        resultsPerPage: 5
      }
    };

    const beforeAutocompleteResultsCall = (options, next) => {
      // Remove sort_direction and sort_field
      // eslint-disable-next-line no-unused-vars
      const { sort_direction, sort_field, ...rest } = options;
      return next({
        ...rest,
        // Add group
        group: { field: "title" }
      });
    };

    await subject({ beforeAutocompleteResultsCall, state, queryConfig });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      q: "searchTerm",
      per_page: 5,
      group: { field: "title" }
    });
  });
});

describe("#onResultClick", () => {
  function subject(clickData) {
    const connector = new SiteSearchAPIConnector(params);
    return connector.onResultClick(clickData);
  }

  it("will call the API with the correct body params", async () => {
    const query = "test";
    const documentId = "12345";

    await subject({
      query,
      documentId
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const url = global.fetch.mock.calls[0][0];
    const urlWithoutTimestamp = url.replace(/&t=\d*/, "").replace(/t=\d*&/, "");
    expect(urlWithoutTimestamp).toEqual(
      `https://search-api.swiftype.com/api/v1/public/analytics/pc?engine_key=${engineKey}&q=${query}&doc_id=${documentId}`
    );
  });
});

describe("#onAutocompleteResultClick", () => {
  function subject(clickData) {
    const connector = new SiteSearchAPIConnector(params);
    return connector.onAutocompleteResultClick(clickData);
  }

  it("will call the API with the correct body params", async () => {
    const query = "test";
    const documentId = "12345";

    await subject({
      query,
      documentId
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const url = global.fetch.mock.calls[0][0];
    const urlWithoutTimestamp = url.replace(/&t=\d*/, "").replace(/t=\d*&/, "");
    expect(urlWithoutTimestamp).toEqual(
      `https://search-api.swiftype.com/api/v1/public/analytics/pas?engine_key=${engineKey}&q=${query}&doc_id=${documentId}`
    );
  });
});
