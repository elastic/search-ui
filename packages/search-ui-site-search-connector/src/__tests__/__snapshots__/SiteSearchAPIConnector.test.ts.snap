// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#onAutocomplete will correctly format an API response 1`] = `
Object {
  "autocompletedResults": Array [
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "9000",
      },
      "id": Object {
        "raw": "5beb474f8db2315427beecc7",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:51:11+00:00",
      },
    },
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "0",
      },
      "id": Object {
        "raw": "5beb47b8a1f2532eb09be102",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:52:56+00:00",
      },
    },
  ],
}
`;

exports[`#onSearch will correctly format an API response 1`] = `
Object {
  "facets": Object {
    "states": Array [
      Object {
        "data": Array [
          Object {
            "count": 2,
            "value": "Maine",
          },
        ],
        "field": "states",
        "type": "value",
      },
    ],
  },
  "rawResponse": Object {
    "errors": Object {},
    "info": Object {
      "national-parks": Object {
        "current_page": 1,
        "facets": Object {
          "states": Object {
            "Maine": 2,
          },
        },
        "num_pages": 1,
        "per_page": 20,
        "query": "acadia",
        "total_result_count": 2,
      },
    },
    "record_count": 2,
    "records": Object {
      "national-parks": Array [
        Object {
          "_explanation": null,
          "_index": "5beb4099c9f929367efa5b7d",
          "_score": 0.7048211,
          "_type": "5beb4099c9f929367efa5b7c",
          "_version": null,
          "acres": 49057.36,
          "external_id": "9000",
          "highlight": Object {
            "title": "<em>Acadia</em>",
          },
          "id": "5beb474f8db2315427beecc7",
          "nps_link": "https://www.nps.gov/acad/index.htm",
          "sort": null,
          "states": Array [
            "Maine",
          ],
          "title": "Acadia",
          "updated_at": "2018-11-13T21:51:11+00:00",
        },
        Object {
          "_explanation": null,
          "_index": "5beb4099c9f929367efa5b7d",
          "_score": 0.7048211,
          "_type": "5beb4099c9f929367efa5b7c",
          "_version": null,
          "acres": 49057.36,
          "external_id": "0",
          "highlight": Object {
            "title": "<em>Acadia</em>",
          },
          "id": "5beb47b8a1f2532eb09be102",
          "nps_link": "https://www.nps.gov/acad/index.htm",
          "sort": null,
          "states": Array [
            "Maine",
          ],
          "title": "Acadia",
          "updated_at": "2018-11-13T21:52:56+00:00",
        },
      ],
    },
  },
  "requestId": "",
  "results": Array [
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "9000",
      },
      "id": Object {
        "raw": "5beb474f8db2315427beecc7",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:51:11+00:00",
      },
    },
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "0",
      },
      "id": Object {
        "raw": "5beb47b8a1f2532eb09be102",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:52:56+00:00",
      },
    },
  ],
  "totalPages": 1,
  "totalResults": 2,
}
`;
